public/app/features/explore/utils/decorators.test.ts

```diff
@@ -86,450 +84,337 @@ const createExplorePanelData = (args: Partial<ExplorePanelData>): ExplorePanelDa
-                [6, 300, '100000000', 'third'],
-              ],
-              type: 'table',
-            })
-          );
-          expect(theResult.fields[0].name).toEqual('value');
-          expect(theResult.fields[1].name).toEqual('time');
-          expect(theResult.fields[2].name).toEqual('tsNs');
-          expect(theResult.fields[3].name).toEqual('message');
-          expect(theResult.fields[1].display).not.toBeNull();
-          expect(theResult.length).toBe(3);
-        },
-        done,
-      });
-    });
-
-    it('should do join transform if all series are timeseries', done => {
-      const tableFrames = [
-        toDataFrame({
-          name: 'A-series',
-          refId: 'A',
-          fields: [
-            { name: 'Time', type: FieldType.time, values: [100, 200, 300] },
-            { name: 'A-series', type: FieldType.number, values: [4, 5, 6] },
-          ],
-        }),
-        toDataFrame({
-          name: 'B-series',
-          refId: 'B',
-          fields: [
-            { name: 'Time', type: FieldType.time, values: [100, 200, 300] },
-            { name: 'B-series', type: FieldType.number, values: [4, 5, 6] },
-          ],
-        }),
-      ];
-      const panelData = createExplorePanelData({ tableFrames });
-
-      observableTester().subscribeAndExpectOnNext({
-        observable: of(panelData).pipe(decorateWithTableResult()),
-        expect: panelData => {
-          const result = panelData.tableResult;
-
-          expect(result?.fields[0].name).toBe('Time');
-          expect(result?.fields[1].name).toBe('A-series');
-          expect(result?.fields[2].name).toBe('B-series');
-          expect(result?.fields[0].values.toArray()).toEqual([100, 200, 300]);
-          expect(result?.fields[1].values.toArray()).toEqual([4, 5, 6]);
-          expect(result?.fields[2].values.toArray()).toEqual([4, 5, 6]);
-        },
-        done,
-      });
-    });
-
-    it('should not override fields display property when filled', done => {
-      const tableFrames = [
-        toDataFrame({
-          name: 'A-series',
-          refId: 'A',
-          fields: [{ name: 'Text', type: FieldType.string, values: ['someText'] }],
-        }),
-      ];
-      const displayFunctionMock = jest.fn();
-      tableFrames[0].fields[0].display = displayFunctionMock;
-
-      const panelData = createExplorePanelData({ tableFrames });
-
-      observableTester().subscribeAndExpectOnNext({
-        observable: of(panelData).pipe(decorateWithTableResult()),
-        expect: panelData => {
-          const data = panelData.tableResult;
-          expect(data?.fields[0].display).toBe(displayFunctionMock);
-        },
-        done,
-      });
-    });
+  it('should process table type dataFrame', async () => {
+    const { table, emptyTable } = getTestContext();
+    const panelData = createExplorePanelData({ tableFrames: [table, emptyTable] });
+    const panelResult = await decorateWithTableResult(panelData).toPromise();
```

Comment:
Moving away from OperatorFunctions introduces `toPromise` in tests which is in my point of view is a hack.

Reply:
Because it's an observable and you're treating it as a Promise.

Reply:
You can transform between those quite easily and in the test you are interested in end result and wan't to see what it is. You don't care here about particular way it is executed here because its 1 input and one observable output. Also for example the `subscribeAndExpectOnNext` which was used before would not handle multiple emissions well because if there were multiple values you still have single expect function and so you don't test different values as they are executed, you would have to have an array of expect functions and probably test if there are more values than expects to make sure all is tested. In that case just collecting the results, waiting for complete and check all the results is more straightforward and you did not loose any information.

This is a bit different from testing an observable pipeline where you have multiple input streams and need to check timing of the events. There handling it with manually mocking the timers is tedious and for example the marble tests make more sense but this is much more simpler case.